### DISCUSSION TITLE
Defining the problem

### DISCUSSION THREAD

**Discussion by lokitkhemka** [TLCFEM/suanPan#238] [General]
I am trying to create a plane-strain problem with quad-elements and monotonic loading. I have nodes that are fixed and nodes that allow only vertical displacements. How can I define the element-type, for the following element line, for example:
```
element 224 374 375 438 437
```
where 224 is the element id and 374, 375, 438 and 437 are the node ids. Is the element type CP4?

Also, how can  I define the fixed and roller nodes (nodes that have only vertical displacements)?

I was going through `triaxial_compression_of_sand` example, but I don't understand the table given in the file `monotonic`

```
0 0
1 1
10 6
```

How can I create an equivalent table for increase of load factor from `0.0` to `1.0` from time `0` to `100`, for example?

Thank you in advance for any help you can provide.

**Comment by TLCFEM** [comment]
1. For [CP4](https://tlcfem.github.io/suanPan-manual/latest/Library/Element/Membrane/Plane/CP4/), you do `element CP4 224 374 375 438 437 material_tag`, leave the thickness as 1 is okay, plane strain is controlled by the material used.
2. Use [PlaneStrain](https://tlcfem.github.io/suanPan-manual/latest/Library/Material/Wrapper/PlaneStrain/) to wrap any 3D material into a plane strain one that can be used with membrane elements.
3. Define [boundary conditions](https://tlcfem.github.io/suanPan-manual/latest/Collection/Define/bc/) using `fix` and `fix2`, fixing both DoF 1 and DoF 2 to have a fixed node, fixing only DoF 1 to allow vertical displacement (if DoF 1 is horizontal axis).
4. A [Tabular](https://tlcfem.github.io/suanPan-manual/latest/Library/Amplitude/Universal/Tabular/) amplitude shall contain a table with two columns which defines the curve.

Check those pages and see which part still cannot be understood.

**Reply by lokitkhemka** [reply]
Hi, I am still not clear about the boundary condition. For example, I have node ids `40 41` that are fixed in all three DoFs and suppose that I have node ids `175 176`, that are free only to have vertical displacement, how would I write them in model scripts?

**Reply by TLCFEM** [reply]
Why would it have 3 DOFs in a 2D problem using CP4?

```
fix 1 1 44 45 175 176 ! fix 1st DOF of nodes 44 45 175 176
fix 2 2 44 45 ! fix 2nd DOF of nodes 44 45
```
Assuming 2 DOFs $u_x$ and $u_y$, nodes 44 45 are fully fixed, nodes 175 176 are rollers.

**Reply by lokitkhemka** [reply]
Certain beam elements also have angular dofs, so I was thinking along that angle. Either way, thank you so much for this.

**Comment by lokitkhemka** [comment]
Hi,

I have few questions regarding the Drucker Prager criterion as implemented in suanpan

In the cookbook the yield surface is defined as $\sqrt{J_2} + \eta p - \xi c $, where $\xi$ and $\eta$ are the parameters. The flow surface is defined $\sqrt{J_2} + \bar{\eta} p $. So, in the input file, if I set `eta_y` and `eta_f`, will I have the associative flow rule?

Also, comparing the yield equation with the equation in the wikipedia [link](https://en.wikipedia.org/wiki/Drucker%E2%80%93Prager_yield_criterion): $\sqrt{J_2} - A - B I_1$, where $A = \frac{6 c \cos \phi}{\sqrt{3} (3 - \sin \phi)}$ and $B =  \frac{2 \cos \phi}{\sqrt{3} (3 - \sin \phi)}$. I can use this equation to compute both $\xi$ and $\eta$ in the form of $\phi$ and $c$ (usually $c$ and $\phi$ are the parameters available to me). Is this approach appropriate? I don't know much about constitutive models, so that is why I am asking to you. However, both the equations need a reference pressure (or hydrostatic pressure) and I see a no way to define it in suanPan ([link to the suanpan doc page](https://tlcfem.github.io/suanPan-manual/latest/Library/Material/Material3D/DruckerPrager/BilinearDP/). I am trying to define Drucker Prager material with no hardening and associative flow rule.

Thank you so much.

**Reply by TLCFEM** [reply]
1. Yes, the associative plasticity is achieved by setting $\eta_y=\eta_f$.
2. There are many different ways to compute $\eta$, the wiki page is correct, furthermore, see [this](https://tlcfem.github.io/suanPan-manual/latest/Library/Material/Material3D/DruckerPrager/NonlinearDruckerPrager/) page.
3. The hydrostatic pressure, $p$ ($I_1$ in wiki page), is computed and updated with the changes of stress, so it is **not** a model parameter. You most likely need to pick a proper friction angle $\phi$ and an initial cohesion $c$.
4. The same model is also documented in [this](https://onlinelibrary.wiley.com/doi/book/10.1002/9780470694626) book, if you have access to, you can check it out.

In general you always want to assign some hardening, say for example 1%, for numerical stability.
A pure plastic response yields zero stiffness, causing convergence issues.

**Reply by lokitkhemka** [reply]
For plane strain, should I use the equations from plain strain fitting from this [link](https://tlcfem.github.io/suanPan-manual/latest/Library/Material/Material3D/DruckerPrager/NonlinearDruckerPrager/)?

**Reply by TLCFEM** [reply]
I think so.
But I won't be too concerned if I have no access to experimental data and am not fitting some curves.
If it is proportional loading, it is always possible to choose a proper friction angle to "predict" the desired capacity.

**Reply by lokitkhemka** [reply]
Ok. Can I calculate the values from outer curve to create a 3d nonlinear dp material model and then define plane strain from there? Because I am trying to run a theoretical model, not fit from experimental values.

**Reply by TLCFEM** [reply]
Not a problem as long as the input is valid.

**Reply by lokitkhemka** [reply]
```
gamma = 20000
E = 1.0E8
c = 6983.24
phi = 57.38
xi = 1.032399
eta = 1.13188
```
```
material BilinearDP 1 1.0e8 0.3 1.13188 1.13188 1.032399 6983.24 0.01 2038.736
material PlainStrain 2 1
```
Is the above material definition valid?

**Reply by TLCFEM** [reply]
What unit system are you using? SI (m)?
Assuming 100 MPa for E, it is relatively large for soil.
The initial cohesion, 6.98 kPa, appears to be too small, some sort of sandy soil?
The density, which is not very important in static analysis, 2038 kg/m^3, seems to be too heavy but still okay.
I guess it looks fine in general but you may want to double check the modulus and cohesion.
And, maybe, using SI (mm) is better.

**Reply by lokitkhemka** [reply]
Yeah, the soil is sandy. I have another question, how can I apply load factor in the analysis? Load factor goes from 0 to 1 in time 0 to say, 50. I am not applying any displacement to the model , just the applying the gravity load.

**Reply by TLCFEM** [reply]
Check [load](https://tlcfem.github.io/suanPan-manual/latest/Collection/Define/load/).

**Reply by lokitkhemka** [reply]
I am not super clear about `magnitude` (should it be unit weight, for body weight?), `dof` tag and `element` tag? Do I have to list all the elements where the force is applied?

**Reply by lokitkhemka** [reply]
I think I understood about the `dof` tag. It must be `1` if the force is in `x` direction and `2` if the force is in `y` direction.

**Reply by TLCFEM** [reply]
1. _The true load magnitude is the product of nominal magnitude and amplitude._ So you do whatever you like, you can assign the specific weight to magnitude and then attach a function that attains unity within the desired analysis duration as the amplitude. Or you can assign unity to load magnitude and define an amplitude that reaches the specific weight within the desired analysis duration. Either way, the product of two will be the specific weight, and the desired body force will be applied.
2. Yes only explicitly assigned elements will be accounted for.
3. DoF understanding is correct. Similarly assigning a moment to a 2D beam involves typically DoF 3, if the DoF is defined as $u_x,u_y,\theta$.

**Reply by lokitkhemka** [reply]
I managed to get the output. However, when I tried to replicate the results from the problem given [here](https://github.com/xcfem/xc_examples/blob/master/geotechnics/slope_stability_analysis/slope_stability_analysis_example_01.py), the behavior is matching, but results have a lot of difference.
[elements.txt](https://github.com/user-attachments/files/18213241/elements.txt)
[nodes.txt](https://github.com/user-attachments/files/18213242/nodes.txt)
[SlopeStabilityInput.txt](https://github.com/user-attachments/files/18213245/SlopeStabilityInput.txt)
These are the input files for suanPan.

The output for the original python program are as follows:

![msrdc_YKPu6xetIA](https://github.com/user-attachments/assets/8deb1acd-4f2d-4712-b8b0-478792b7afb6)
![msrdc_zHnxjQUwE9](https://github.com/user-attachments/assets/5890783e-ff5f-420c-9f09-7b92f7febdc7)

**Reply by TLCFEM** [reply]
I checked it, overall it looks good.
But there is one thing.
As shown [here](https://tlcfem.github.io/suanPan-manual/latest/Collection/Define/load/#__codelineno-2-4), the default ramp amplitude is used when amplitude tag is 0.
And [here](https://tlcfem.github.io/suanPan-manual/latest/Library/Amplitude/Universal/Linear/), you find the amplitude of a ramp function is unity.
So even you have assigned an analysis time of 100s, from t=1s to t=100s, within this whole 99s the load stays constant, which is the load magnitude -20k times 1, which is -20k.
Is this the value intended? If not, you may just change the load magnitude and use a unit analysis step with a small step size.

```text
load bodyforce 1 0 <a proper value> 2 ... ! element tags are removed here for brevity
step static 1 1 ! use unity analysis time
set fixed_step_size 1
set ini_step_size 1E-2 ! with 100 time steps
```

Otherwise you define a linear amplitude.

Also as I said previously, the elastic modulus is relatively large since SI (m) is used.
A strong convergence criterion is used for local iteration, which may fail in this case.
I have relaxed the criterion in this commit: 8ffa2be6c0926f84adc1f911cd3c4a37eb409a4e

With which the following ultmate status can be obtained with the maximum load 0.281*-200k=-56k of whatever load unit used in the original setup.
![example sliding line](https://github.com/user-attachments/assets/cd94c2bd-6121-4cb1-9f88-62c4bf4d6af6)

Not sure if it makes much sense though, if not, some parameters need to be tweaked.

**Reply by lokitkhemka** [reply]
The intention is to increase the value of load factor from 0 to 1 (i.e., the load increases from 0 to -20k) in time 0 to 100. I am now, using `tabular` data as below:

```
0 0
100 1
```

Is this valid now?

**Reply by TLCFEM** [reply]
Yes, this serves the purpose.

**Reply by lokitkhemka** [reply]
The relaxed convergence test works well. But, I am not sure why? Perhaps, it has something to do with high `E` value, but I am not sure why. Can you explain it to me? This is one question that has been bugging me for almost a decade now (why do local iteration fail to converge and what can be done to rectify it?).

The values are still not matching. However, the behavior is exactly the same. The plots of `U1` and `U2` are below:
![2024-12-21_12-29](https://github.com/user-attachments/assets/33405f22-7eb5-4881-8296-1736f22ffbe0)
![2024-12-21_12-29_1](https://github.com/user-attachments/assets/8ee8f98e-8d9a-4d5f-925f-c4989b9a343e)

**Reply by TLCFEM** [reply]
Without knowing the details of one or both implementations, it's not likely to get the exact values.
For example, consider two bilinear hardening models with 0.01 and 0.008 hardening ratios, to achieve the same level of load, the required displacement can differ vastly, depending on how high the load is above initial yielding.

I'm not too concerned. But if you do, what you can do is to check the implementations and create a single element model under uniaxial loading and compare the results.

In terms of local convergence, the absolute magnitude of stress/modulus would affect the absolute magnitude of residual, especially in those models where quantities and functions are not defined in a dimensionless space.
The convergence criterion shall check one of absolute and relative errors, a universal better criterion shall check both.
However, both could fail due to different reasons. For absolute error, if the initial absolute magnitude of residual is large, it may not be able to converge to the prescribed tolerance. For relative error, if the "condition number" of the algorithm/formulation is bad, meaning it involves some computation that loses precision, it may also fail to converge to a prescribed tolerance.

A few things can be done:
1. use a proper unit system so that numbers are nicer to deal with,
2. optimise the formulation/implementation,
3. use a smarter convergence criterion that may handle more cases.

**Reply by lokitkhemka** [reply]
Anyways, I guess it is good enough. Can you use it as an example in suanPan docs? I have attached all the input files.
[SlopeStabilityInput.txt](https://github.com/user-attachments/files/18219027/SlopeStabilityInput.txt) 

[elements.txt](https://github.com/user-attachments/files/18219024/elements.txt)
[load_factor.txt](https://github.com/user-attachments/files/18219025/load_factor.txt)
[nodes.txt](https://github.com/user-attachments/files/18219026/nodes.txt)

Thank  you so much for your explanation of the why local convergence can fail.

**Reply by TLCFEM** [reply]
I'd like to do so, but I have no time to draft an example based on the inputs.

Ideally, I'd like to have the following elements.
1. A markdown file that contains the example.
2. A zip file contains all input files, and one can unzip can run the main script with no additional setup, `.txt` files shall be renamed to `.sp` files so that syntax highlighting would work in Sublime Text and VS Code.
3. Present the context, explain how the model is established (selection of parameters, etc.), and discuss engineering practice implications (comparing with for example what Bishop's method gives).

I am able to do maybe 50% of the above tasks at most.
Are you interested in writing up something, as you clearly know more in this field?
Otherwise I presume it is not going to happen anytime soon.

**Reply by lokitkhemka** [reply]
I will spend some time writing it up and maybe you can check, if it works. I am hoping to write it, maybe after the new year, because I have some other work until new years or even until 8th of Jan (I have IELTS on 8th). But, I will try to get working on it as much as I can, in the meantime too.

**Reply by lokitkhemka** [reply]
[Slope Stability Analysis in Geotechnical Engineering.zip](https://github.com/user-attachments/files/18262242/Slope.Stability.Analysis.in.Geotechnical.Engineering.zip)
I have written the documentation for this. Can you please clean it up? It has been a while since I have had to write anything related to geotechnical engineering. I have put the folder in the `Example` of the docs and hence, the paths of the links are relative to that.

**Reply by TLCFEM** [reply]
I made this available here: https://tlcfem.gitbook.io/suanpan-manual/example/geotechnical/slope-analysis
You can find the corresponding page in other deployments.
Revisions if any can be made later.

**Reply by lokitkhemka** [reply]
Thank you for allowing me to contributing.

**Comment by TLCFEM** [comment]
BTW, feel free to revise the documentation you have read and clarify any existing confusion. I believe it can be improved, and feedback from a reader's perspective would be beneficial.

**Reply by lokitkhemka** [reply]
I will go through the documentation to see if I can do that too.
